### ISSUE TITLE
Firestore emulator sometimes fails on data export when stopped

### ISSUE THREAD

**Issue by madmacc** [firebase/firebase-tools#3092]
**Describe your environment:**

  * Operating System version:  Windows 10
  * Browser version: Chrome  88.0.4324.104
  * Firebase SDK version: 8.0.0
  * Firebase Product: Firestore local emulator

**Problem:**
When I stop my local firestore emulator most of the time it exports the data fine.  However sometimes is has an error and it deletes my export files so I lose all my data.  Nothing seems to appear in the logs.
I am using:
firebase emulators:start --import=emulator/emulator-export-data --export-on-exit=emulator/emulator-export-data

Then:
Ctrl+C
to stop the emulator.

I get the following in the terminal:

> i  emulators: Received SIGINT (Ctrl-C) for the first time. Starting a clean shutdown.
> i  emulators: Please wait for a clean shutdown or send the SIGINT (Ctrl-C) signal again to stop right now.
> i  Automatically exporting data using --export-on-exit "emulator/emulator-export-data" please wait for the export to finish...
> i  Found running emulator hub for project hivepass-dev at http://localhost:4401
> i  Deleting directory R:\hivepass\HivePassApp\emulator\emulator-export-data\firestore_export
> i  Exporting data to: R:\hivepass\HivePassApp\emulator\emulator-export-data
> !  Export request failed, see emulator logs for more information.
> !  Automatic export to "emulator/emulator-export-data" failed, going to exit now...
> i  emulators: Shutting down emulators.
> i  ui: Stopping Emulator UI
> !  Emulator UI has exited upon receiving signal: SIGINT
> i  hosting: Stopping Hosting Emulator
> i  firestore: Stopping Firestore Emulator
> !  Firestore Emulator has exited upon receiving signal: SIGINT
> i  auth: Stopping Authentication Emulator
> i  hub: Stopping emulator hub
> i  logging: Stopping Logging Emulator

There is not much in the firestore-debug.log in the root:

> INFO: Detected non-HTTP/2 connection.
> Jan 28, 2021 12:46:37 PM io.gapi.emulators.netty.HttpVersionRoutingHandler channelRead
> INFO: Detected non-HTTP/2 connection.
> Jan 28, 2021 12:46:38 PM io.gapi.emulators.netty.HttpVersionRoutingHandler channelRead
> INFO: Detected non-HTTP/2 connection.
> Jan 28, 2021 12:46:38 PM io.gapi.emulators.netty.HttpVersionRoutingHandler channelRead
> INFO: Detected non-HTTP/2 connection.
> Jan 28, 2021 12:46:39 PM io.gapi.emulators.netty.HttpVersionRoutingHandler channelRead
> INFO: Detected non-HTTP/2 connection.
> Jan 28, 2021 12:46:39 PM io.gapi.emulators.netty.HttpVersionRoutingHandler channelRead
> INFO: Detected non-HTTP/2 connection.

**Comment by samhorlbeck** [comment]
Hi @madmacc, thanks for the report.

Would you mind expanding a bit about the steps you're taking before seeing this issue? For example, how much data is being saved? What SDKs are you using? Are you using any of the other emulators in conjunction with Firestore?

**Comment by madmacc** [comment]
Sure.
The data is only 130k.
I am running the following emulators.  Firestore, Auth, Functions
However I am running the functions emulator as a separate instance as the functions folder is in a different project/location and I can't figure out how to run all of them with one command.

I am using v9.2.1 of the firebase-tools.

Let me know if you need any more information.

**Comment by samhorlbeck** [comment]
What is your general use case? Can you discern any particular action that reliably causes the data to not be saved correctly? Basically I'm just trying to get a sense of how we could start to debug this issue and potentially reproduce it.

**Comment by madmacc** [comment]
Unfortunately not.  I'm sure there is some reason but it seems random to me at this stage.
I can run it loading the previously exported data.  Add some more data and then stop it by hitting ctrl+c in the terminal.
It will sometimes export ok and sometimes not.  It only deletes the firestore_export folder not the firebase-export-metadata.json file.

At the moment when it is successful I manually copy the exported data to a safe location so it is not deleted. 

I was hoping the logs would show something but they don't unless I am looking in the wrong place.  Root of the project firestore-debug.log.

I am just testing an web application locally using the firebase js sdk and angularfire.

**Comment by madmacc** [comment]
I just tried starting and stopping it about 10 times in a row and it worked every time.
I tried adding some data, adding data from my app and even putting the computer into sleep mode and coming back out.

So not sure what it is but it is either time related or caused by something else.

**Comment by madmacc** [comment]
I just had the issue again after leaving my computer for a couple of hours in sleep mode.  Slightly different error this time.

Log is attached.  Looks the same.  I'm not sure if this is correct log file as it doesn't have much interesting in it.

> i  emulators: Received SIGINT (Ctrl-C) for the first time. Starting a clean shutdown.
> i  emulators: Please wait for a clean shutdown or send the SIGINT (Ctrl-C) signal again to stop right now.
> i  Automatically exporting data using --export-on-exit "emulator/emulator-export-data" please wait for the export to finish...
> !  Did not find any running emulators for project hivepass-dev.
> !  Automatic export to "emulator/emulator-export-data" failed, going to exit now...
> i  emulators: Shutting down emulators.
> i  ui: Stopping Emulator UI
> !  Emulator UI has exited upon receiving signal: SIGINT
> i  hosting: Stopping Hosting Emulator
> i  firestore: Stopping Firestore Emulator
> !  Firestore Emulator has exited upon receiving signal: SIGINT
> i  auth: Stopping Authentication Emulator
> i  hub: Stopping emulator hub
> !  hub: Error stopping emulator hub
> i  logging: Stopping Logging Emulator
[firestore-debug.log](https://github.com/firebase/firebase-tools/files/5891350/firestore-debug.log)

**Comment by yuchenshi** [comment]
> ! Did not find any running emulators for project hivepass-dev.

The Emulator Suite writes a temporary file to keep track of the running instances, and looks like that file was deleted somehow before you terminated the suite. For example, Windows may delete those temporary files to reclaim disk space (especially if disk space is low), or it may be some other cleanup tools installed on your dev machine.

Right now, `--export-on-exit` relies on that file, but it doesn't have to -- I've created b/179048252 to track making it more reliable. Thanks again for the bug report!

**Comment by madmacc** [comment]
Interesting.  Could it be because I am running two instances of the emulator as stated above?
One for functions (as my functions folder is in a different project directory) and one for everything else.

**Comment by yuchenshi** [comment]
Glad we found that out! The export APIs cannot handle two instances running together yet, but the internal bug I've created above should be able to solve this.

In the meantime, have you considered creating symlinks so that your functions can be "part of" the main project directory? They are available on Windows 10 and you can ignore them in `.gitignore` (or similar) if you'd like to.

**Comment by madmacc** [comment]
> In the meantime, have you considered creating symlinks so that your functions can be "part of" the main project directory? They are available on Windows 10 and you can ignore them in `.gitignore` (or similar) if you'd like to.

This is a good idea.  I have not used them before but have had them on my todo list, I will give it a try.  I tried a shortcut which of course did not work!

It must be common scenario for people that have multiple apps in their firebase project.  Not sure why I am the only one who has brought it up.

**Comment by madmacc** [comment]
Actually after some more testing the first issue with the deleted export data files is due to the two instances of the emulator running.
But only if I start the firestore emulator first then the functions emulator and then stop the firestore emulator while the functions emulator is still running.

**Comment by JoeEarly** [comment]
Am seeing this also on Mac, every other time my export files are either corrupted or missing data. 

Will comeback with the detail

**Comment by carloschida** [comment]
On macOS 11.1 with `firebase-tools@9.3.0` I'm seeing something of the kind:

On WebStorm 2020.3.2 when using the "Run" button on `emulators:start`:

```json
// package.json

    "emulators:start": "firebase emulators:start --project=\"local-emulators\" --import=\"./a\" --export-on-exit=\"./a\""
```

The result is:
```shell
/Users/carlos.chida/.nvm/versions/node/v12.13.0/bin/node /usr/local/Cellar/yarn/1.22.5/libexec/bin/yarn.js run emulators:start
yarn run v1.22.10
$ firebase emulators:start --project="local-emulators" --import="./a" --export-on-exit="./a"
...
┌─────────────────────────────────────────────────────────────┐
│ ✔  All emulators ready! It is now safe to connect your app. │
│ i  View Emulator UI at http://localhost:4000                │
└─────────────────────────────────────────────────────────────┘

### STOPPED THE PROCESS VIA WEBSTORM ###

i  emulators: Received SIGINT (Ctrl-C) for the first time. Starting a clean shutdown.
i  emulators: Please wait for a clean shutdown or send the SIGINT (Ctrl-C) signal again to stop right now.
i  Automatically exporting data using --export-on-exit "./a" please wait for the export to finish...
⚠  Emulator UI has exited upon receiving signal: SIGINT
i  Found running emulator hub for project local-emulators at http://localhost:4400
i  Creating export directory /Users/carlos.chida/WebstormProjects/veriphi/functions/a
i  Exporting data to: /Users/carlos.chida/WebstormProjects/veriphi/functions/a
```

But **only** the folder gets created without any contents. I'm assuming that WebStorm is not giving the process enough time after sending the interrupt signal.

On the other hand, when running `yarn emulators:start`, it works without problems. Nonetheless, it looks like it consumes much time after receiving the signal:
```shell
yarn emulators:start
yarn run v1.22.10
$ firebase emulators:start --project="local-emulators" --import="./a" --export-on-exit="./a"
...
### PRESSED ctrl-C 
i  emulators: Received SIGINT (Ctrl-C) for the first time. Starting a clean shutdown.
i  emulators: Please wait for a clean shutdown or send the SIGINT (Ctrl-C) signal again to stop right now.
i  Automatically exporting data using --export-on-exit "./import-export-data" please wait for the export to finish...
i  Found running emulator hub for project local-emulators at http://localhost:4400                                                                                                                                     
i  Creating export directory /Users/carlos.chida/WebstormProjects/veriphi/functions/import-export-data
i  Exporting data to: /Users/carlos.chida/WebstormProjects/veriphi/functions/import-export-data
i  emulators: Received export request. Exporting data to /Users/carlos.chida/WebstormProjects/veriphi/functions/import-export-data.
$ ### NEW PROMPT!
➜ ✔  emulators: Export complete.
✔  Export complete
i  emulators: Shutting down emulators.
i  ui: Stopping Emulator UI
⚠  Emulator UI has exited upon receiving signal: SIGINT
i  functions: Stopping Functions Emulator
i  database: Stopping Database Emulator
i  firestore: Stopping Firestore Emulator
i  pubsub: Stopping Pub/Sub Emulator
⚠  functions: The Cloud Firestore emulator is not running, so calls to Firestore will affect production.
i  auth: Stopping Authentication Emulator
i  hub: Stopping emulator hub
$ ### ANOTHER NEW PROMPT
```

**Comment by yuchenshi** [comment]
@carloschida  Thanks for the report! Exporting may take a few seconds and if WebStorm isn't giving the emulator enough time to do the work, there isn't a ton we can do here. We suggest running `emulators:start` in a terminal and you may find the built-in WebStorm terminal emulator better than the "Run" button for your case. Anyway, feel free to contact WebStorm regarding the issue.

**Comment by wceolin** [comment]
Not sure if it's related to this issue but a similar error is happening to a co-worker:

```
emulators: Export failed: Server Error. read ECONNRESET
```

Our emulator script:

```sh
firebase --emulators:exec --import="./exports/current" --export-on-exit --ui "next dev"
```

It's very hard to reproduce since it happens only sometimes. I haven't been able to reproduce it myself but I was wondering if you have any ideas of what this could be or what I could investigate.

Another thing I'm finding weird is that we have duplicated shutting down operations:

```
emulators: Received export request. Exporting data to /exports/current.
✔  emulators: Export complete.
✔  emulators: Export complete.
✔  Export complete
i  emulators: Shutting down emulators.
i  ui: Stopping Emulator UI
✔  Export complete
i  emulators: Shutting down emulators.
i  ui: Stopping Emulator UI
⚠  Emulator UI has exited upon receiving signal: SIGINT
i  functions: Stopping Functions Emulator
i  functions: Stopping Functions Emulator
i  database: Stopping Database Emulator
i  database: Stopping Database Emulator
i  firestore: Stopping Firestore Emulator
i  firestore: Stopping Firestore Emulator
i  pubsub: Stopping Pub/Sub Emulator
i  pubsub: Stopping Pub/Sub Emulator
i  auth: Stopping Authentication Emulator
i  auth: Stopping Authentication Emulator
i  hub: Stopping emulator hub
i  hub: Stopping emulator hub
i  logging: Stopping Logging Emulator
i  logging: Stopping Logging Emulator
```

You can all operation happen twice. However, that seems to be related to the `next dev` script used by [Next.js](https://nextjs.org/). That doesn't happen with other scripts (i.e. Jest in watch mode).

**Comment by yuchenshi** [comment]
@wceolin This is an interesting case and would you be able to get the `firebase-debug.log` for this one? That may be helpful for troubleshooting what caused everything to happen twice.

**Comment by yuchenshi** [comment]
@wceolin I appreciate the help, but we need the `firebase-debug.log` (created by CLI), not `firestore-`.

**Comment by wceolin** [comment]
@yuchenshi Sorry. I wasn't able to find a `firebase-debug.log` file. There's only `database-debug`, `firestore-debug`, `pubsub-debug`, and `ui-debug`. Not sure if this helps but the `database-debug.log` had the following warning:

```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by io.netty.util.internal.ReflectionUtil (file:/Users/$username/.cache/firebase/emulators/firebase-database-emulator-v4.7.2.jar) to field sun.nio.ch.SelectorImpl.selectedKeys
WARNING: Please consider reporting this to the maintainers of io.netty.util.internal.ReflectionUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
19:10:08.230 [NamespaceSystem-akka.actor.default-dispatcher-5] INFO akka.event.slf4j.Slf4jLogger - Slf4jLogger started
19:10:08.387 [main] INFO com.firebase.server.forge.App$ - Listening at 0.0.0.0:9000
19:35:26.787 [Thread-0] INFO com.firebase.server.forge.App$ - Attempting graceful shutdown.
19:35:26.799 [NamespaceSystem-akka.actor.default-dispatcher-10] INFO com.firebase.core.namespace.Terminator$Terminator - 1 actors left to terminate: project-name
19:35:26.805 [NamespaceSystem-akka.actor.default-dispatcher-10] INFO com.firebase.core.namespace.NamespaceActor - Stopping namespace actor for project-name
19:35:26.806 [NamespaceSystem-akka.actor.default-dispatcher-10] INFO com.firebase.core.namespace.NamespaceActor - Gauges removed for project-name
19:35:26.810 [Thread-0] INFO com.firebase.server.forge.App$ - Graceful shutdown complete.
```

**Comment by muellerdberlin** [comment]
> On macOS 11.1 with `firebase-tools@9.3.0` I'm seeing something of the kind:
> 
> On WebStorm 2020.3.2 when using the "Run" button on `emulators:start`:


As mentioned, in Webstorm I could handle it by using the Webstorm terminal and not the Run Configuration. However, that's strange.

**Comment by iplaksiy** [comment]
This happens intermittently to me as well but only if I leave the emulators running for a few days and the machine goes through various wake and sleep cycles during that time.

**Comment by wceolin** [comment]
Yeah, we haven't seen this issue for almost two weeks now.

**Comment by madmacc** [comment]
I just came back to a project I had not run in the emulator for a few weeks and the data was gone again.  Not sure how or when it happened.  I assume it just didn't export correctly when I shut it down the last time.
Can be quite frustrating as I need to regularly make a manual backup of the data in case this happens.

**Comment by wceolin** [comment]
I'm still not able to consistently reproduce the "export failed" error but this issue is happening to 3 co-workers at the moment. In all cases, there's no `firebase-debug.log` file. Enabling the `--debug` flag, we got the following information:

```sh
emulators: Export failed: Server Error. read ECONNRESET
[2021-07-15T15:00:42.106Z] <<< HTTP RESPONSE 500 {"x-powered-by":"Express","vary":"Origin","content-type":"application/json; charset=utf-8","content-length":"43","etag":"W/\"2b-PTnaVGW4PHglIRlUWNl1/Oxt2Ys\"","date":"Thu, 15 Jul 2021 15:00:42 GMT","connection":"keep-alive","keep-alive":"timeout=5"}
[2021-07-15T15:00:42.107Z] <<< HTTP RESPONSE BODY Server Error. read ECONNRESET {"message":"Server Error. read ECONNRESET"}
⚠  Export request failed, see emulator logs for more information. {"name":"FirebaseError","children":[],"exit":1,"message":"Export request failed, see emulator logs for more information.","original":{"name":"FirebaseError","children":[],"context":{"body":{"message":"Server Error. read ECONNRESET","error":{"message":"Unknown Error"}},"response":{"statusCode":500,"body":{"message":"Server Error. read ECONNRESET","error":{"message":"Unknown Error"}},"headers":{"x-powered-by":"Express","vary":"Origin","content-type":"application/json; charset=utf-8","content-length":"43","etag":"W/\"2b-PTnaVGW4PHglIRlUWNl1/Oxt2Ys\"","date":"Thu, 15 Jul 2021 15:00:42 GMT","connection":"keep-alive","keep-alive":"timeout=5"},"request":{"uri":{"protocol":"http:","slashes":true,"auth":null,"host":"localhost:4400","port":"4400","hostname":"localhost","hash":null,"search":null,"query":null,"pathname":"/_admin/export","path":"/_admin/export","href":"http://localhost:4400/_admin/export"},"method":"POST"}}},"exit":2,"message":"HTTP Error: 500, Unknown Error","status":500},"status":500}
⚠  Automatic export to "./firebase-sdk/exports/current" failed, going to exit now...
```

Sometimes, we also get this:

```sh
logging: Stopping Logging Emulator
(node:3153) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'exit' of undefined
    at /home/pbeeson/sb/remote-control/node_modules/firebase-tools/lib/command.js:92:40
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
(node:3153) 
```

That seems to come [from this line](https://github.com/firebase/firebase-tools/blob/master/src/command.ts#L201) suggesting that maybe the error is `undefined`?

These export failed issues seem to happening only when using the `emulators:exec`. We haven't seen them with `emulators:start` AFAIK.

I'm going to try digging more into the `firebase-tools` codebase to see if I can have a better understanding of what might be going on here but I'm posting the information above in case someone more familiar with the codebase has any ideas.

**Comment by wceolin** [comment]
I'm still not able to consistently reproduce the "export failed" error but this issue is happening to 3 co-workers at the moment. In all cases, there's no `firebase-debug.log` file. Enabling the `--debug` flag, we got the following information:

```sh
emulators: Export failed: Server Error. read ECONNRESET
[2021-07-15T15:00:42.106Z] <<< HTTP RESPONSE 500 {"x-powered-by":"Express","vary":"Origin","content-type":"application/json; charset=utf-8","content-length":"43","etag":"W/\"2b-PTnaVGW4PHglIRlUWNl1/Oxt2Ys\"","date":"Thu, 15 Jul 2021 15:00:42 GMT","connection":"keep-alive","keep-alive":"timeout=5"}
[2021-07-15T15:00:42.107Z] <<< HTTP RESPONSE BODY Server Error. read ECONNRESET {"message":"Server Error. read ECONNRESET"}
⚠  Export request failed, see emulator logs for more information. {"name":"FirebaseError","children":[],"exit":1,"message":"Export request failed, see emulator logs for more information.","original":{"name":"FirebaseError","children":[],"context":{"body":{"message":"Server Error. read ECONNRESET","error":{"message":"Unknown Error"}},"response":{"statusCode":500,"body":{"message":"Server Error. read ECONNRESET","error":{"message":"Unknown Error"}},"headers":{"x-powered-by":"Express","vary":"Origin","content-type":"application/json; charset=utf-8","content-length":"43","etag":"W/\"2b-PTnaVGW4PHglIRlUWNl1/Oxt2Ys\"","date":"Thu, 15 Jul 2021 15:00:42 GMT","connection":"keep-alive","keep-alive":"timeout=5"},"request":{"uri":{"protocol":"http:","slashes":true,"auth":null,"host":"localhost:4400","port":"4400","hostname":"localhost","hash":null,"search":null,"query":null,"pathname":"/_admin/export","path":"/_admin/export","href":"http://localhost:4400/_admin/export"},"method":"POST"}}},"exit":2,"message":"HTTP Error: 500, Unknown Error","status":500},"status":500}
⚠  Automatic export to "./firebase-sdk/exports/current" failed, going to exit now...
```

Sometimes, we also get this:

```sh
logging: Stopping Logging Emulator
(node:3153) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'exit' of undefined
    at /my-project/node_modules/firebase-tools/lib/command.js:92:40
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
(node:3153) 
```

That seems to come [from this line](https://github.com/firebase/firebase-tools/blob/master/src/command.ts#L201) suggesting that maybe the error is `undefined`?

These export failed issues seem to happening only when using the `emulators:exec`. We haven't seen them with `emulators:start` AFAIK.

I'm going to try digging more into the `firebase-tools` codebase to see if I can have a better understanding of what might be going on here but I'm posting the information above in case someone more familiar with the codebase has any ideas.

**Comment by michaelyfan** [comment]
Can confirm that I'm also consistently getting this error. Start the emulators in one console window with `firebase emulators:start`, make some data in the emulated Firestore UI, run `firebase emulators:export sample_data_directory_here` in another console window, and then I get the `Error: Did not find any running emulators for project [project-name].` message. What's going on? Am using the most up to date Firebase CLI.

**Comment by scotteau** [comment]
Can confirm the problem occurs when running firebase emulator within webstorm run window, but works fine in terminal.

**Comment by wceolin** [comment]
I bought a new MacBook Air this week and the export failed error is happening every time on my new machine (I'm using the terminal, btw), which is weird because it's basically the same setup I have in another machine (a Mac Mini M1) where it doesn't happen.

Initially, I thought this could be because I was using Node 14 on my new machine and Node 12 in my old one (that's the only difference I could spot) but that's not the case since a co-worker told me she's having the same issue using Node 12.

This also doesn't seem to be OS specific. This is happening to two co-workers using Ubuntu, to another one using MacOS with an Intel processor, and now with me using MacOS with an M1 processor.

**Comment by 50an6xy06r6n** [comment]
@yuchenshi is there a way to recover from this state? I seem to run into it occasionally when I make changes to my code that crash the function (I'm running tsc in watch mode).

**Comment by MorenoMdz** [comment]
We have this issue as well when we leave the emulator running for long times, is there a way to recover from this? The file the emulator writes can be saved somehow to be reused later when this happen in case it gets removed?

**Comment by Drew-Daniels** [comment]
I have been having the same issue on Windows 10 (Firebase Tools version 10.7.1). It happens regardless of if the emulator has been running for a long time or just a minute or two, and when the amount of data is minimal (i.e., one collection with one document with one field). I try exporting emulator data to a local folder directory named `data` but when I shut down I get the following exported to my root directory:

![image](https://user-images.githubusercontent.com/62859552/164904386-d8d8605e-d5dc-4b8b-9ed1-b88063ac9f23.png)


When I run `firebase emulators:start --export-on-exit=data`, this is my PowerShell output:
```
i  emulators: Starting emulators: auth, firestore, database, hosting, pubsub, storage
!  emulators: It seems that you are running multiple instances of the emulator suite for project drew-daniels-wheres-waldo. This may result in unexpected behavior.
i  firestore: Firestore Emulator logging to firestore-debug.log
i  database: Database Emulator logging to database-debug.log
i  pubsub: Pub/Sub Emulator logging to pubsub-debug.log
i  hosting: Serving hosting files from: build
+  hosting: Local server: http://localhost:5000
i  ui: Emulator UI logging to ui-debug.log

┌─────────────────────────────────────────────────────────────┐
│ ✔  All emulators ready! It is now safe to connect your app. │
│ i  View Emulator UI at http://localhost:4000                │
└─────────────────────────────────────────────────────────────┘

┌────────────────┬────────────────┬─────────────────────────────────┐
│ Emulator       │ Host:Port      │ View in Emulator UI             │
├────────────────┼────────────────┼─────────────────────────────────┤
│ Authentication │ localhost:9099 │ http://localhost:4000/auth      │
├────────────────┼────────────────┼─────────────────────────────────┤
│ Firestore      │ localhost:8080 │ http://localhost:4000/firestore │
├────────────────┼────────────────┼─────────────────────────────────┤
│ Database       │ localhost:9000 │ http://localhost:4000/database  │
├────────────────┼────────────────┼─────────────────────────────────┤
│ Hosting        │ localhost:5000 │ n/a                             │
├────────────────┼────────────────┼─────────────────────────────────┤
│ Pub/Sub        │ localhost:8085 │ n/a                             │
├────────────────┼────────────────┼─────────────────────────────────┤
│ Storage        │ localhost:9199 │ http://localhost:4000/storage   │
└────────────────┴────────────────┴─────────────────────────────────┘
  Emulator Hub running at localhost:4400
  Other reserved ports: 4500

Issues? Report them at https://github.com/firebase/firebase-tools/issues and attach the *-debug.log files.
```
Everything appears to run fine, but once I `ctrl+C` it appears that the emulator does not shut down gracefully:
```
i  emulators: Received SIGINT (Ctrl-C) for the first time. Starting a clean shutdown.
i  emulators: Please wait for a clean shutdown or send the SIGINT (Ctrl-C) signal again to stop right now.
i  Automatically exporting data using --export-on-exit "data" please wait for the export to finish...

Terminate batch job (Y/N)? Error: Storage Emulator Rules runtime exited unexpectedly.
i  Found running emulator hub for project (my-project-name) at http://localhost:4400
i  Exporting data to: C:\Users\MyUserName\my\folder\project\data
i  emulators: Received export request. Exporting data to C:\Users\MyUserName\my\folder\project\data.
```

What I have tried:

- Changing the relative file paths from ./data to data
- Ensuring all java.exe instances are closed (via task manager)
- Restarting my computer

Made an SO post on this yesterday:
https://stackoverflow.com/questions/71972842/firebase-emulator-export-on-exit-not-working-properly-windows-10?noredirect=1#comment127180514_71972842

**Comment by nayfin** [comment]
My co-worker was having the exact same issue as described above by @Drew-Daniels , and I wasn't. He was also seeing 2 java terminal's open, but I wasn't. After upgrading to the latest `firebase-tools` I started getting a warning that I need to upgrade Java to at least v11 (not sure what I had initially or why it was so old). So I upgrade Java to v18, now I am seeing the 2 java terminals open when the emulator starts, and my emulator data is getting exported to a folder at my project root instead of the  export location. Before upgrading Java it seemed to create this root folder for a few seconds on shutdown, then move the contents over, now it seems like something is interrupting the process halfway through.  I'm tagging @yuchenshi and @weixifan  since this seems to at give some insight into a change that can initiate the issue. Let me know if I should create a new issue or if you think this is related.

**Comment by abakucz** [comment]
I do experience this issue as well. 

And it also started after I updated Java to 18 (and also upgraded NodeJS 14 from 12).

**Comment by nayfin** [comment]
I just noticed that @Drew-Daniels was getting the `!  emulators: It seems that you are running multiple instances of the emulator suite...` error. Which seems like it's common among people having this issue of the data not being exported where expected. I am not getting error on initialization, but I'm still having the data export issue.

No error on initialization
![image](https://user-images.githubusercontent.com/15178439/166801194-53d63a63-bd1b-44e9-9c38-993b6711beb8.png)

But data is still being created in hashed folder instead of in the specified folder:
![image](https://user-images.githubusercontent.com/15178439/166801850-b66d7b2a-21ff-412d-99d1-a7ba392c9228.png)

**Comment by abderahmaneBasil** [comment]
Same problem here but with a different context. I dockerized the emulators and wanted to export the data in a volume when container exits but it fails every time with the following error :
![image](https://user-images.githubusercontent.com/73698016/168428223-4e8c70aa-9640-4d9d-93c8-1678682b27e2.png)
![image](https://user-images.githubusercontent.com/73698016/168428268-cf19f6f9-4e83-456b-93c2-cfb8e44c0e27.png)
![image](https://user-images.githubusercontent.com/73698016/168428304-62609b9e-2463-4f4f-8071-2fa79e5e1374.png)

**Comment by dereekb** [comment]
On the Docker issue, I ran into the same thing, but it unable to save was a “problem” with Docker rather than the emulator.

Your Firebase data folder can’t be the mapped folder, you need to map a parent folder and try and export to a child folder.

Here’s my mappings and usage:

https://github.com/dereekb/dbx-components/blob/develop/docker-compose.yml

https://github.com/dereekb/dbx-components/blob/7a5e05a27ab83976ba7dec6430477c8ecd458416/workspace.json#L685


Still, I’ll occasionally have the emulator dump an empty folder in the root folder, and I’m not sure exactly what causes that.

**Comment by abderahmaneBasil** [comment]
Thanks @dereekb 🙌
It was as you wrote : I mapped to a parent folder and it works perfectly now.

**Comment by nayfin** [comment]
Can someone with a working instance please lookup and post your versions for the following dependencies: Java, node, firebase, firebase-tools? Mapping to a parent folder didn't work for me and we're not using docker. This is the emulator start command we run: 
`firebase emulators:start --import=./emulator/export-data --export-on-exit=./emulator/export-data`

**Comment by ntorrey** [comment]
Same problem and behavior as @nayfin 
Windows 11
java 17.0.1
node 16.13.1
firebase 9.8.1
firebase-tools 11.0.1

**Comment by Darkade** [comment]
I'm guessing this is related but running `firebase emulators:export .`  just deleted my whole repo after failing to run. I understand if this isn't the syntax the command expected but deleting the whole repo?!

**Comment by yuchenshi** [comment]
@Darkade We've tried to make the command not do that. Please make sure you're using the latest Firebase CLI version. If that still happens, please see https://github.com/firebase/firebase-tools/issues/3962#issuecomment-1128169529 for a longer discussion and feel free to provide any relevant information there.

Friendly reminder to keep this discussion on export failures. If you have any information to add on this front, even just OS / Java / Node versions, feel free to add a comment to help us pinpoint this. (Right now, it looks related to some Java 17/18 changes on specific platforms -- more info appreciated.) If you have a different problem, don't hesitate to open a new issue and we'll help you there.

**Comment by Darkade** [comment]
I mean, the export did failed before what happened @yuchenshi, but I get you, thanks. If it helps these are the versions I'm using

```
$ java --version
openjdk 11.0.15 2022-04-19
OpenJDK Runtime Environment Homebrew (build 11.0.15+0)
OpenJDK 64-Bit Server VM Homebrew (build 11.0.15+0, mixed mode)
$ node --version
v14.19.3
$ firebase --version
11.0.1
$  arch
arm64
$ sw_vers
ProductName:	macOS
ProductVersion:	12.4
BuildVersion:	21F79
```

**Comment by rhuanbarros** [comment]
I have the same issue

**Comment by rhuanbarros** [comment]
I just recreated the project in WSL Ubuntu, installed java 11, and everything worked just fine.

**Comment by nesimtunc** [comment]
Experiencing the same issue. None of those solution works for me.

**Comment by kiptoomm** [comment]
Same problems here. My setup:

$ java -version
java version "18.0.1.1" 2022-04-22
Java(TM) SE Runtime Environment (build 18.0.1.1+2-6)
Java HotSpot(TM) 64-Bit Server VM (build 18.0.1.1+2-6, mixed mode, sharing)

$ firebase -V
11.1.0

$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.15.5
BuildVersion:	19F101

$ node --version
v16.4.2

**Comment by leonelqf** [comment]
Same here...

**Comment by CodyTolene** [comment]
Same here. 

Node v16.14.2
Firebase v11.2.1

Script ran is `firebase emulators:start --only functions,firestore:indexes,firestore --import=./.data --export-on-exit` in the terminal and then initiating a clean shutdown with `CTRL + C`.

Resulting error:
```
!  emulators: Export failed: EPERM: operation not permitted, rename 'firebase-export-1657573896802aUSBs6' -> 'C:\TestFunctions\.data'
! Export request failed, see emulator logs for more information.
```

There is nothing logged regarding this issue where stated within "firestore-debug.log".  Also the data successfully exports but it's just not to the "./.data/" as in the script but to the path "./firebase-export-1657573896802aUSBs6/" instead.

Please advise. Thank you

**Comment by yuchenshi** [comment]
Thanks for everyone's input. Looks like we've seen a new trend of permission errors when trying to rename the temporary directory to the final destination. I'm not 100% sure about the permission model of Windows and Linux here, but could it be because that another process (maybe a Java process that failed to exit) still has data files open in the directory? One hypothesis may be that newer Java versions take longer to close files / does not close file handles cleanly on exit.

**Comment by dereekb** [comment]
I can reproduce different behavior in my [dbx-components project](https://github.com/dereekb/dbx-components) in a couple of ways. I start the Docker container using `./serve-server.sh`, which begins my server instance. It will load content from `./.firebase/data/emulators` (mapped to `/root/data/emulators` in the container) where it is set to import from.

For some extra background, I'm running two functions in parallel, one uses Nx's build/watch feature to rebuild the project when a change occurs, and I've got [entr](https://eradman.com/entrproject/) set up to watch for changes to the `dist/apps/demo-api` folder where Nx dumps the new build, then runs `firebase --project=default emulators:start --import=/root/data/emulators --export-on-exit`. 

Whenever `entr` sees the dist folder change change, then it sends a shutdown signal (a single SIGTERM) to the emulators.

Here's the logs output for that:

```
┌─────────────────────────────────────────────────────────────┐
│ ✔  All emulators ready! It is now safe to connect your app. │
│ i  View Emulator UI at http://0.0.0.0:9900                  │
└─────────────────────────────────────────────────────────────┘

┌────────────────┬──────────────┬───────────────────────────────┐
│ Emulator       │ Host:Port    │ View in Emulator UI           │
├────────────────┼──────────────┼───────────────────────────────┤
│ Authentication │ 0.0.0.0:9903 │ http://0.0.0.0:9900/auth      │
├────────────────┼──────────────┼───────────────────────────────┤
│ Functions      │ 0.0.0.0:9902 │ http://0.0.0.0:9900/functions │
├────────────────┼──────────────┼───────────────────────────────┤
│ Firestore      │ 0.0.0.0:9904 │ http://0.0.0.0:9900/firestore │
├────────────────┼──────────────┼───────────────────────────────┤
│ Hosting        │ 0.0.0.0:9901 │ n/a                           │
├────────────────┼──────────────┼───────────────────────────────┤
│ Pub/Sub        │ 0.0.0.0:9905 │ n/a                           │
├────────────────┼──────────────┼───────────────────────────────┤
│ Storage        │ 0.0.0.0:9906 │ http://0.0.0.0:9900/storage   │
└────────────────┴──────────────┴───────────────────────────────┘
  Emulator Hub running at localhost:4400
  Other reserved ports: 4500

Issues? Report them at https://github.com/firebase/firebase-tools/issues and attach the *-debug.log files.
 
chunk (runtime: main) main.js (main) 704 KiB [entry] [rendered] // <----- Nx finished another build
webpack compiled successfully (045a918f5004ecc7)
 
i  emulators: Received SIGTERM for the first time. Starting a clean shutdown.  // <--- entr sends shutdown to emulators
i  emulators: Please wait for a clean shutdown or send the SIGTERM signal again to stop right now.
i  Automatically exporting data using --export-on-exit "/root/data/emulators" please wait for the export to finish...
i  Found running emulator hub for project dereekb-components at http://localhost:4400
i  Exporting data to: /root/data/emulators
i  emulators: Received export request. Exporting data to /root/data/emulators.

Waiting for ports 9900-9906 to be available.                        // <--- wait-for-ports.sh run to wait for emulators to shut down
Ports still in use. Waiting for ports to close... PIDs: 1422
1450
1477
1526

Error: Storage Emulator Rules runtime exited unexpectedly.   // <--- weird crash that I'm assuming causes the rest of the emulators to freeze up

Ports still in use. Waiting for ports to close... PIDs: 1450
1477
1526
Ports still in use. Waiting for ports to close... PIDs: 1450
1477
1526
Ports still in use. Waiting for ports to close... PIDs: 1450
1477
1526
Ports still in use. Waiting for ports to close... PIDs: 1450
1477
1526
Preparing to forcefully kill ports...  // <--- wait-for-ports.sh kills the remaining 3 processes
```

It seems like rules emulator fails to close properly, then causes the others to seize up and not finish their own exports and close until my [wait-for-ports.sh](https://github.com/dereekb/dbx-components/blob/develop/wait-for-ports.sh) script forcefully closes them.

Here's a list of the failed exports that have accumulated as an example of the structure being saved:

<img width="493" alt="Screen Shot 2022-07-12 at 1 28 40 AM" src="https://user-images.githubusercontent.com/3586580/178426753-c088a316-c899-40d9-9c77-221f7b391e8d.png">

A new folder is created each time I make a backend change that Nx captures.

Here's the logs for this test:

[logs.txt](https://github.com/firebase/firebase-tools/files/9090262/logs.txt)

**Comment by bradintheusa** [comment]
If anyone needs more examples of it not working I'm happy to help, It's driving me crazy and taking my productivity to zero.

**Comment by ulises-castro** [comment]
For my case I forgot to indicate in --export-on-exit the same path I use for importing data
```bash
firebase emulators:start --import ../../path-to-firebase-data  --export-on-exit ../../path-to-firebase-data 
```

**Comment by bradintheusa** [comment]
The rollback suggested by @abakucz works for me

```
    "firebase-admin": "^10.1.0",
    "firebase-functions": "^3.20.1",

```

**Comment by bradintheusa** [comment]
Make sure your Java install is a full install as per the install instructions https://jdk.java.net/18/

**Comment by LuisAlejandro** [comment]
This happens to me on a dockerized firebase emulation with the following versions:

```
        "firebase-admin": "^11.0.1",
        "firebase-functions": "^3.22.0",
```

**Comment by danbaechtold** [comment]
In my case the problem is related to **Vite.js**:

When I start the emulators and then run Vite (dev server), and then stop the emulators, the error occurs.
But when I start the emulators without running Vite, or when I stop Vite first, it works without any problems.

I assume the way Vite works it somehow locks emulator files/folders (I've named the import/export directory "emulators").

Btw both processes are started with NPM scripts in VS Code terminal, configured in package.json like this:
```
"emulators": "firebase emulators:start --import=emulators --export-on-exit",
"dev": "vite"
```

**Comment by christhompsongoogle** [comment]
Hi [danbaechtold](https://github.com/danbaechtold), if you use a different export directory not inside the current working directory (like /tmp), does it still fail the same way?

**Comment by danbaechtold** [comment]
Hi @christhompsongoogle, yes, it still fails when I move it into the parent directory. Weird enough, even when I set `--export-on-exit=../emulators`, it writes the fallback export `firebase-export-1665...` not to the parent, but to the current directory..

**Comment by lennartguth** [comment]
I'm having the exact same issue as @danbaechtold .
It writes all exported data to `firebase-export-*random-number*` instead to the folder I specified with `--export-on-exit=./firebase-export`. The folder `firebase-export` was created by the command but it is empty.

**Comment by julesmqzsma** [comment]
Same issue as danbaechtold and lennartguth, here's my package json:
`{
  "name": "functions",
  "scripts": {
    "lint": "eslint --ext .js,.ts . --fix",
    "build": "tsc",
    "build:watch": "tsc --watch",
    "serve": "firebase emulators:start --import=./firebase_data --export-on-exit --only functions,firestore,auth,hosting:fleet-sistema-api",
    "shell": "npm run build && firebase functions:shell",
    "start": "npm run shell",
    "deploy": "firebase deploy --only functions",
    "logs": "firebase functions:log"
  },
  "engines": {
    "node": "16"
  },
  "main": "lib/index.js",
  "dependencies": {
    "cors": "^2.8.5",
    "express": "^4.18.1",
    "firebase-admin": "^10.0.2",
    "firebase-functions": "^4.0.1",
    "uuid": "^9.0.0"
  },
  "devDependencies": {
    "@types/uuid": "^8.3.4",
    "@typescript-eslint/eslint-plugin": "^5.12.0",
    "@typescript-eslint/parser": "^5.12.0",
    "eslint": "^8.9.0",
    "eslint-config-google": "^0.14.0",
    "eslint-plugin-import": "^2.25.4",
    "firebase-functions-test": "^0.2.0",
    "typescript": "^4.5.4"
  },
  "private": true
}
`

Additionally, I am in a windows 11 machine and I do have a couple of discs, my workspace is saved in D while my npm was installed in C via nvm. Do you think this may be an issue?

**Comment by Darkness-2** [comment]
> 

Thanks for pointing this out! This was my issue when attempting to export firebase emulators. Vite seems to lock-down the entire project folder in some way. Closing Vite first solves the issue.

**Comment by Proibito** [comment]
same here with vite

**Comment by dhust** [comment]
> I'm guessing this is related but running `firebase emulators:export .` just deleted my whole repo after failing to run. I understand if this isn't the syntax the command expected but deleting the whole repo?!

The same thing just happened to me a few minutes ago. Lost my entire project! Please tell me there's a way to get it all back. Add why does an export delete stuff. Why is not just a file that gets made?

**Comment by bleuscyther** [comment]
> > I'm guessing this is related but running `firebase emulators:export .` just deleted my whole repo after failing to run. I understand if this isn't the syntax the command expected but deleting the whole repo?!
> 
> The same thing just happened to me a few minutes ago. Lost my entire project! Please tell me there's a way to get it all back. Add why does an export delete stuff. Why is not just a file that gets made?

If you are using inteliJ you can "undelete" by checking the local history  of the parent directory

![image](https://user-images.githubusercontent.com/2999487/204063197-e8b75f26-d534-4dcb-921e-d0666dfd0494.png)

**Comment by Giuliano-Stravini** [comment]
Hey yo!

I had the same issue today and i solved this problem executing the command specifing what modules of Firebase i want, like:

`firebase emulators:start --only firestore,auth,functions,pubsub,storage --import=./firebase-export --export-on-exit`

If we don't use --only <modules>, like the example, always when exit, the emulator will export in differents folders.

This workaround worked to me.

**Comment by janwuesten** [comment]
I have the same issue on Windows 11 (MacOS is fine) but for me storing the export data on a different drive than the Emulator is running on fixes the issue for me.
So if I run the emulator on Disk 1, I can store my data on Disk 2 but not Disk 1. If I run my emulator on Disk 2, I can store the data on Disk 1 but not Disk 2.

As a workaround I created a virtual disk (VHD) inside the windows disk manager and mount it as F: to store my exported data onto it. Not the best solution as I would like to store the data inside the project directory, but maybe it still helps someone until a fix is provided.

`firebase emulators:start --import=F:/projectname --export-on-exit=F:/projectname`

**Comment by RichardPragnell** [comment]
> I'm having the exact same issue as @danbaechtold . It writes all exported data to `firebase-export-*random-number*` instead to the folder I specified with `--export-on-exit=./firebase-export`. The folder `firebase-export` was created by the command but it is empty.

Any luck with this? I'm also having the same issue.

**Comment by arisAlexis** [comment]
please fix exporting I have this error plus another error in another issue. Clearly export does not work correctly for a lot of people and needs more testing.

**Comment by MorenoMdz** [comment]
> please fix exporting I have this error plus another error in another issue. Clearly export does not work correctly for a lot of people and needs more testing.

Indeed. We plan on fixing this in our project by moving of Firestore.

**Comment by arisAlexis** [comment]
I opened a new issue here with multiple bugs also including this one.  #5404

**Comment by bmcbarron** [comment]
Based on my experimentation, it looks like Vite dev server holds a lock on folders that contain sub-folders (at least on Windows). Filed bug report https://github.com/vitejs/vite/issues/11938

**Comment by sugoidesune** [comment]
So I run `"firebase emulators:start --import firestore_temp --export-on-exit"`

1. It deletes `firestore_temp`
2. It creates it's own folders
3. It keeps creating new ones, never importing data.
![image](https://user-images.githubusercontent.com/20317691/219700144-1b101091-ceea-4ee2-8bc1-d64118523ff4.png)


I only have the functions and emulator running in there...

**Comment by ahmednfwela** [comment]
@sugoidesune a temporary workaround is to change save path to somewhere on `c://` drive, e.g. `c://firebase/exports`

**Comment by sugoidesune** [comment]
> @sugoidesune a temporary workaround is to change save path to somewhere on `c:/` drive, e.g. `c:/firebase/exports`

Did you manage to do relative paths. `../folder`  didn't work for me either.

**Comment by TheoOliveira** [comment]
> @sugoidesune a temporary workaround is to change save path to somewhere on `c:/` drive, e.g. `c:/firebase/exports`

in linux would that work?

**Comment by bmcbarron** [comment]
If, like me, your symptoms are caused by the Vite dev server holding locks on nested folders (which can prevent the emulator from renaming its snapshots on Windows), the workaround I'm using is to set `server: { watch: { usePolling: true }}` in `vite.config.ts.`  That changes the implementation of the directory watching and I haven't had issues ever since.

**Comment by YKakdas** [comment]
Creating a VHD and using it as the export folder worked for me. No matter what I tried, nothing worked when I only had 1 disk drive partition.

**Comment by joe528** [comment]
The export always succeeds if I launch the emulator a few minutes/hours ago.

But if I leave the emulator running for an extensive time (maybe 1-2 days) on my Mac, the export fails always. 

Hope this can get fixed. I just lost data again today since I forgot about this issue.

**Comment by dennismuench** [comment]
It is happening no matter what (MacOS). For me this feature never worked. I have this exact problem since I first started experimenting with Firebase years ago. I had this problem on Windows, Mac and with every framework. Even if I just run a static SPA through the hosting emulator (no active dev server). Compiling the comments so far it looks like it is firebase-tools own file change detection mechanism that is locking things up...

**Comment by SKempin** [comment]
Running `firebase emulators:export ./tmp` results in the following log error for me. The dir is created but is empty. Any further progress with this?

`⚠  emulators: Export failed: ENOENT: no such file or directory, mkdtemp 'firebase-export-1691618521143XXXXXX'`

**Comment by christhompsongoogle** [comment]
[SKempin](https://github.com/SKempin), the emulators create the export in a temp directory and then copy them over once completed:

https://github.com/firebase/firebase-tools/blob/40dd38f8c8ce8095ea18abe623fb5102184ee239/src/emulator/hubExport.ts#L134

The mkdtemp creation is failing for some reason, inside of /tmp

See: https://www.geeksforgeeks.org/node-js-fs-mkdtempsync-method/

Are you running this inside a container or with permissions that would prevent the firebase CLI from creating a temp directory in /tmp?

**Comment by SKempin** [comment]
@christhompsongoogle thanks, no not running in a container and have also tried running with `sudo`, and multiple different directory paths. Same result each time.

**Comment by FAR5HID** [comment]
> So I run `"firebase emulators:start --import firestore_temp --export-on-exit"`
> 
> 1. It deletes `firestore_temp`
> 2. It creates it's own folders
> 3. It keeps creating new ones, never importing data.
>    ![image](https://user-images.githubusercontent.com/20317691/219700144-1b101091-ceea-4ee2-8bc1-d64118523ff4.png)
> 
> I only have the functions and emulator running in there...

Same situation when running in container

**Comment by HirukaChansilu** [comment]
> If, like me, your symptoms are caused by the Vite dev server holding locks on nested folders (which can prevent the emulator from renaming its snapshots on Windows), the workaround I'm using is to set `server: { watch: { usePolling: true }}` in `vite.config.ts.` That changes the implementation of the directory watching and I haven't had issues ever since.

Thanks, This worked also for me.

**Comment by remillc** [comment]
No comments since october 2023, am I missing something?

This problem is major for the developer experience.

**Comment by yanir21** [comment]
> No comments since october 2023, am I missing something?
> 
> This problem is major for the developer experience.
> 
> For me, changing the path to `C:\firebase-temp-data` did the trick.

Still not working for me. Does anyone have a different workaround?

**Comment by SKempin** [comment]
@yanir21 I am running this npm script now to run the emulator and it is successfully exporting on every stop, as long as I stop using `ctrl + c`

`"serve": "npm run build && firebase emulators:start --import emulator_data --export-on-exit emulator_data"`

I have a `emulator_data` dir at root level of the project.

**Comment by yanir21** [comment]
@SKempin Still no luck:

```shell
...
i  Exporting data to: C:\Users\USER\Desktop\functions\emulator_data
i  emulators: Received export request. Exporting data to C:\Users\USER\Desktop\functions\emulator_data.
!  emulators: Export failed: EPERM: operation not permitted, rename 'firebase-export-1707209884206CzcLW7' -> 'C:\Users\USER\Desktop\functions\emulator_data'
!  Export request failed, see emulator logs for more information.
!  Automatic export to "emulator_data" failed, going to exit now...
i  emulators: Shutting down emulators.
...
```

**Comment by SKempin** [comment]
@yanir21 can you push a repo for me to try this?

**Comment by mognutecnologia** [comment]
Hello, in my case it fails exactly as described by Mainul Islam Farshid,
but I found out that my "running server" process was holding the lock of
the whole directory. In someone's case there could be even others holding
the lock.

So what I have been doing until firebase team solves this:

1 - Before starting the emulator, I check if last export succeeded, if not,
I manually rename the failed exported directory to "firebase-export-data"
2 - Start emulator making sure the export name matches the import argument
when starting emulator: eg: " --import firebase-export-data"

> firebase emulators:start --import ./firebase-export-data --export-on-exit
> ./firebase-export-data --inspect-functions

3 - Work
4 - Before exiting, I stop serving, compiling, and everything else that may
be holding the lock *before *I exit the emulator.
5 - Making sure the emulator is the last running process, I kill it with *one
single* ctrl+c, and wait for it to export (can take a few seconds, so don't
hit several ctrl+C).
6 - I check the directory, if I still have my "firebase-export-data"
folder, if it fails it will change to something like:
"firebase-export-1707209884206CzcLW7".
In this case I manually rename it to "firebase-export-data", to make sure I
don't forget next time I will start the emulator (step 1).

Needs a bit of extra attention, but works for me.

**Comment by yanir21** [comment]
I finally managed to solve this 🎉

The root directory (e.g `/project` ) holds the `firebase.json` file, but I tried to run the emulators from `/projects/functions` because that's where my `package.json` npm scripts are. 

Running `emulators:start` from the root project directory successfully exports the data to the specified folder, and also shuts down gracefully (which is another problem I had).

**Comment by lssilveira11** [comment]
@yanir21 solution worked for me too, thanks!

**Comment by reimertz** [comment]
@yanir21 you da real mvp!

**Comment by ngblaylock** [comment]
> Our issue was relying on `concurrently` to spawn multiple processes (used to build typescript).
> 
> The issue is twofold;
> 
> * When sigterming `concurrently` it sends multiple sigterm downwards to processes it has initiated.
> * If `firebase-tools` recieve more than one `CTRL-C`, it force-exits without shutting down gracefully.
> 
> This has the side-effect of import / export not gracefully exporting and explains why different solutions such as disabling `storage`, running the command in different folders etc. might result in the import / export being able to finish before being force-quit.
> 
> I hope this can help you all!

This seems to answer my issue. I changed my npm script to be this: 

<img width="723" alt="Screenshot 2024-02-24 at 1 52 54 PM" src="https://github.com/firebase/firebase-tools/assets/16627408/618024c7-a119-4a98-8066-ea2d697ae92b">

Now all my logs can still be in the same terminal, and using <kbd>command + c</kbd> closes all of them and successfully completes the export. The logs are not quite as pretty, but my export works now which is more important.

**Comment by mbpictures** [comment]
Still having an error during export:
```
i  Creating export directory B:\Projects\DigitalChoirLeader\packages\firebase\firebase-export
i  Exporting data to: B:\Projects\DigitalChoirLeader\packages\firebase\firebase-export
i  emulators: Received export request. Exporting data to firebase\firebase-export.
!  emulators: Export failed: EPERM: operation not permitted, rename 'firebase\firebase-export-1709550643283fOGVYX' -> 'firebase\firebase-export'
!  Export request failed, see emulator logs for more information.
!  Automatic export to "./firebase-export" failed, going to exit now...
i  emulators: Shutting down emulators.
i  ui: Stopping Emulator UI
!  Emulator UI has exited upon receiving signal: SIGINT
i  firestore: Stopping Firestore Emulator
!  Firestore Emulator has exited upon receiving signal: SIGINT
```

Even trying to manually rename the specified folder failed with the error, that the contents of the directory are still in use by some program. Do you have an idea on how to fix this issue?

**Comment by JFGHT** [comment]
> Our issue was relying on `concurrently` to spawn multiple processes (used to build typescript).
> 
> The issue is twofold;
> 
> * When sigterming `concurrently` it sends multiple sigterm downwards to processes it has initiated.
> * If `firebase-tools` recieve more than one `CTRL-C`, it force-exits without shutting down gracefully.
> 
> This has the side-effect of import / export not gracefully exporting and explains why different solutions such as disabling `storage`, running the command in different folders etc. might result in the import / export being able to finish before being force-quit.
> 
> I hope this can help you all!

This worked.

**Comment by Dhinesh-Dhinesh** [comment]
# Solution for Running Firebase Emulator with Import/Export Outside the `/functions` Folder

If you encounter issues when running the Firebase emulator while inside the `/functions` directory, try executing the command from outside that folder. This approach ensures smoother operation when building and starting the emulator.

## Steps to Run Firebase Emulator

Navigate to the root of your project (outside the `/functions` directory) and use the following command:

```
yarn build && cd .. && firebase emulators:start --import ./emulator-data --export-on-exit ./emulator-data --inspect-functions
```

### Explanation:
- **`yarn build`**: Compiles the TypeScript or JavaScript code in your `/functions` directory.
- **`cd ..`**: Moves to the parent directory (project root).
- **`firebase emulators:start --import ./emulator-data --export-on-exit ./emulator-data --inspect-functions`**: Starts the Firebase emulator, imports data from `./emulator-data`, and exports data on exit to the same location while enabling inspection of functions for debugging.

By running the command from the root directory, you can avoid potential path and permission issues that may arise when executing it directly within the `/functions` folder.

### Note:
Ensure that the `./emulator-data` folder exists in the project root or modify the path as necessary for your setup.

**Comment by rsb-tbg** [comment]
Heres how I'm doing import/export in a container and it has worked well so far.

firebase.Dockerfile
```
FROM google/cloud-sdk:latest

# Install Node.js (required for Firebase CLI)
RUN apt-get update && apt-get install -y curl gnupg && \
    curl -fsSL https://deb.nodesource.com/setup_20.x | bash - && \
    apt-get install -y nodejs && \
    npm install -g firebase-tools && \
    apt-get clean && rm -rf /var/lib/apt/lists/*

# Set default command
CMD ["firebase", "emulators:start", \
"--project=my-project", \
"--config=/data/firebase.json", \
"--import=/data/exports", \
"--export-on-exit=/data/exports", \
"--only=auth,firestore", \
"--debug"]
```

docker-compose.yaml (you don't need to pass your creds if you don't want):
```
services:
  firebase-emulator:
    container_name: firebase-emulator
    platform: linux/amd64
    build:
      context: .
      dockerfile: firebase.Dockerfile
    ports:
      - "9099:9099"
      - "8080:8080"
      - "4000:4000"
    environment:
      - GCP_PROJECT_ID=my-project
      - GOOGLE_APPLICATION_CREDENTIALS=/tmp/gcloud-creds.json
    volumes:
      - ~/.config/gcloud/application_default_credentials.json:/tmp/gcloud-creds.json:ro
      - ./firebase-data:/data
    restart: always
```

and structure of my mounted firebase-data/
```
firebase-data/
├── firebase.json          # your emulator config (ports, rules, etc)
└── exports/
    └── [empty or previous exports…]
```

Hopefully this helps.

**Comment by Vloz** [comment]
Fixe for EPERM on VsCode+Win11 here:
https://github.com/firebase/firebase-tools/issues/1112#issuecomment-3373003369

**Comment by purezen** [comment]
I lost a number of data updates on restart today. 

Is it the same issue? 15.2.1 on MacOS.

**Comment by tamannasgh** [comment]
okay so this worked for me,
first stop your server like vite that is using firebase emulators, and after that you can just stop the emulators using ctrl + c and it successfully exports the data and then kill the terminal and close the running java file.
